# Exploring Chromatin Occupancy with ChIP-seq, CUT&RUN, or CUT&Tag
---------------------------------

## **Contents**

+ [Overview](#overview)
+ [Background](#background)
+ [Before Starting](#before-starting)
+ [Getting Started](#getting-started)
+ [Software Requirements](#software-requirements)
+ [Architecture Design](#architecture-design)
+ [Data](#data)
+ [Funding](#funding)
+ [License for Data](#license-for-data)

## **Overview**
<img src="images/LessonPlan.jpg" width="600" />

## **Background**
There are several reasons to explore the occupancy of proteins on chromatin. ChIP-seq, CUT&RUN, and CUT&Tag represent three high-throughput sequencing approaches to profile chromatin occupancy genome-wide. These approaches have been used to measure histone modifications, RNA polymerases, architectural proteins, and transcription factors, to name a few. These approaches use an antibody to selectively isolate DNA fragments bound by a protein of interest. While these three approaches share some similarities, there are differences that should be taken into consideration when planning experiments and during the analysis of the data. 

### **ChIP-seq**
<img src="images/chipseq.gif" width="600" />

Most ChIP-seq protocols including

## **Before Starting**


## **Getting Started**


## **Software Requirements**


## **Architecture Design**


## **Data**


## **Funding**


## **License for Data**

